### PR TITLE
Add support for setting custom fields

### PR DESCRIPTION
**Pull request by phemmylintry**
implementation for https://github.com/withlogicco/hq/issues/311

### CODE REVIEW

File: src/add-to-project.ts [hunk 1]
```diff
@@ -152,11 +224,37 @@ export async function addToProject(): Promise<void> {
...
+        input: {
+          projectId,
+          itemId,
+          fieldId: key,
+          value,
+        },
+      },
+    )
+
+    core.setOutput('itemId', updateResp.updateProjectV2ItemById.item.id)
```
Comment (phemmylintry) [inline src/add-to-project.ts]: @akalipetis I don't know why, but this block of code does not run

Comment (akalipetis) [reply src/add-to-project.ts]: @phemmylintry when you have the time, please push the code that we debugged together 💪🏽

Comment (phemmylintry) [reply src/add-to-project.ts]: @akalipetis I've pushed the update

File: .licenses/npm/@actions/core.dep.yml [hunk 2]
```diff
@@ -1,20 +1,20 @@
...
+      The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
-    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (akalipetis) [inline .licenses/npm/@actions/core.dep.yml]: Let's avoid changing files that we don't touch and keep the same formatting.

File: .licenses/npm/uuid.dep.yml [hunk 3]
```diff
@@ -3,18 +3,18 @@ name: uuid
...
+      The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
 
-    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+      THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (akalipetis) [inline .licenses/npm/uuid.dep.yml]: Ditto

File: __tests__/add-to-project.test.ts [hunk 4]
```diff
@@ -13,7 +13,7 @@ describe('addToProject', () => {
   beforeEach(() => {
     mockGetInput({
       'project-url': 'https://github.com/orgs/actions/projects/1',
-      'github-token': 'gh_token'
+      'github-token': 'gh_token',
```
Comment (akalipetis) [inline __tests__/add-to-project.test.ts]: Let's keep the same formatting in this file as before.

File: __tests__/add-to-project.test.ts [hunk 5]
```diff
@@ -90,48 +90,96 @@ describe('addToProject', () => {
...
     await addToProject()
 
     expect(outputs.itemId).toEqual('project-item-id')
   })
 
+  test('add an issue with fields product and client', async () => {
+    mockGetInput({
+      'project-url': 'https://github.com/orgs/actions/projects/1',
+      'github-token': 'gh_token',
+      fields: 'product=test-product, client=test-client',
```
Comment (akalipetis) [inline __tests__/add-to-project.test.ts]:
```suggestion
      fields: 'product=test-product\nclient=test-client',
```

Shouldn't fields be one per line?

File: __tests__/add-to-project.test.ts [hunk 6]
```diff
@@ -90,48 +90,96 @@ describe('addToProject', () => {
...
+          addProjectV2ItemById: {
+            item: {
+              id: 'project-item-id',
+            },
+          },
+        },
+      },
+    )
+    await addToProject()
+    expect(outputs.itemId).toEqual('project-item-id')
```
Comment (akalipetis) [inline __tests__/add-to-project.test.ts]: Can we check here that the fields were added correctly? Does the mock allows us to verify a specific call was made?

File: action.yml [hunk 7]
```diff
@@ -16,6 +16,9 @@ inputs:
   label-operator:
     required: false
     description: The behavior of the labels filter, AND to match all labels, OR to match any label, NOT to exclude any listed label (default is OR)
+  fields:
+    required: false
+    description: A json object containing custom fields like product and client
```
Comment (akalipetis) [inline action.yml]: This should not be a JSON object, it should be a multiline input containing `field=value` pairs.

File: dist/index.js [hunk 8]
```diff
@@ -101,7 +101,7 @@ function addToProject() {
       }
     }`, {
             projectOwnerName,
-            projectNumber
+            projectNumber,
```
Comment (akalipetis) [inline dist/index.js]: Ditto

File: jest.config.js [hunk 9]
```diff
@@ -3,7 +3,7 @@ module.exports = {
   moduleFileExtensions: ['js', 'ts'],
   testMatch: ['**/*.test.ts'],
   transform: {
-    '^.+\\.ts$': 'ts-jest'
+    '^.+\\.ts$': 'ts-jest',
   },
-  verbose: true
+  verbose: true,
```
Comment (akalipetis) [inline jest.config.js]: Ditto

File: package-lock.json [hunk 10]
```diff
@@ -663,6 +663,20 @@
       "integrity": "sha512-0hYQ8SB4Db5zvZB4axdMHGwEaQjkZzFjQiN9LVYvIFB2nSUHW9tYpxWriPrWDASIxiaXax83REcLxuSdnGPZtw==",
       "dev": true
     },
+    "node_modules/@cspotcode/source-map-support": {
```
Comment (akalipetis) [inline package-lock.json]: Let's revert all changes here, since we've not updated any dependencies.

File: src/add-to-project.ts [hunk 11]
```diff
@@ -39,6 +59,13 @@ interface ProjectV2AddDraftIssueResponse {
 export async function addToProject(): Promise<void> {
   const projectUrl = core.getInput('project-url', {required: true})
   const ghToken = core.getInput('github-token', {required: true})
+  const fields = core.getInput('fields', {required: true})
```
Comment (akalipetis) [inline src/add-to-project.ts]: This should be a multiline input.

File: src/add-to-project.ts [hunk 12]
```diff
@@ -39,6 +59,13 @@ interface ProjectV2AddDraftIssueResponse {
...
+  const fieldsObj = fields.split('\n').reduce((acc, f) => {
```
Comment (akalipetis) [inline src/add-to-project.ts]: Fields will be already split if we define them as a multiline input[^1]

[^1]: https://github.com/actions/toolkit/blob/1589a5c066fbe34ed1c828e9b81b9f206f9a3a40/packages/core/src/core.ts#L140-L161

File: src/add-to-project.ts [hunk 13]
```diff
@@ -113,6 +140,52 @@ export async function addToProject(): Promise<void> {
...
+      }
+    }`,
+    {
+      projectId,
+    },
+  )
+
+  const projectFields = fieldsResp.node?.fields.nodes ?? []
+
+  const results = await check.checkDictionaryInArray(fieldsObj, projectFields)
```
Comment (akalipetis) [inline src/add-to-project.ts]: This should be moved later on and should only be used if there's at least one field in input - if there are not fields, there's no need to make this call.

File: src/add-to-project.ts [hunk 14]
```diff
@@ -152,11 +226,35 @@ export async function addToProject(): Promise<void> {
...
+        input: {
+          projectId,
+          itemId,
+          fieldId: result.id,
+          value: result.value,
+        },
+      },
+    )
+
+    core.setOutput('itemId', updateResp.updateProjectV2ItemFieldValue.projectV2Item.id)
```
Comment (akalipetis) [inline src/add-to-project.ts]: This is not needed, as this should have already been set as output above.

File: src/check-custom-fields.ts [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export async function checkDictionaryInArray(obj: {[key: string]: string}, arr: []) {
+  const object: ObjectType = obj
+  const array: ArrayType[] = arr
+
+  const result = []
+
+  for (const key in object) {
+    if (object.hasOwnProperty(key)) {
```
Comment (akalipetis) [inline src/check-custom-fields.ts]: Do we really need this one? Iteration should work for all keys without an issue.

File: src/check-custom-fields.ts [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+      const found = array.find(item => item.name.toLowerCase() === key.toLowerCase())
+
+      if (found && found.options) {
+        const optionFound = found.options.find(option => option.name.toLocaleLowerCase() === object[key].toLowerCase())
+        if (optionFound) {
+          const id = found.id
+          const value = optionFound.id
+          result.push({id, value: {singleSelectOptionId: value}})
+        }
+      }
```
Comment (akalipetis) [inline src/check-custom-fields.ts]: We should at least handle number, text and single select fields. Let's leave iterations out for now.

File: src/add-to-project.ts [hunk 17]
```diff
@@ -113,6 +140,52 @@ export async function addToProject(): Promise<void> {
...
+                id
+                name
+                dataType
+                configuration {
+                  iterations {
+                    startDate
+                    id
+                  }
+                }
+              }
```
Comment (akalipetis) [inline src/add-to-project.ts]: We should leave iterations out for now and we'll figure them out at a later PR.

File: README.md [hunk 18]
```diff
@@ -142,8 +142,9 @@ Using these events ensure that a given issue or pull request, in the workflow's
 
 - add the newly created PAT as a repository secret, this secret will be referenced by the [github-token input](#github-token)
   _See [Encrypted secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for more information_
-  
+
 ## Setting a specific status or column name to the project item
+
```
Comment (parisk) [inline README.md]: Let's revert these changes, as they are out of scope.

File: src/add-to-project.ts [hunk 19]
```diff
@@ -79,7 +101,7 @@ export async function addToProject(): Promise<void> {
 
   if (!urlMatch) {
     throw new Error(
-      `Invalid project URL: ${projectUrl}. Project URL should match the format https://github.com/<orgs-or-users>/<ownerName>/projects/<projectNumber>`
+      `Invalid project URL: ${projectUrl}. Project URL should match the format https://github.com/<orgs-or-users>/<ownerName>/projects/<projectNumber>`,
```
Comment (parisk) [inline src/add-to-project.ts]: Let's revert this change, as it is out of scope.

File: src/add-to-project.ts [hunk 20]
```diff
@@ -103,8 +125,8 @@ export async function addToProject(): Promise<void> {
     }`,
     {
       projectOwnerName,
-      projectNumber
-    }
+      projectNumber,
+    },
```
Comment (parisk) [inline src/add-to-project.ts]: ditto: https://github.com/withlogicco/add-to-project/pull/1/files#diff-e3ec74736882b1cbe4eced52e9c1cea7812a3428c168dfddf7cb15dbed18d561R104

File: src/add-to-project.ts [hunk 21]
```diff
@@ -130,12 +184,13 @@ export async function addToProject(): Promise<void> {
       {
         input: {
           projectId,
-          contentId
-        }
-      }
+          contentId,
+        },
+      },
```
Comment (parisk) [inline src/add-to-project.ts]: ditto: https://github.com/withlogicco/add-to-project/pull/1/files#diff-e3ec74736882b1cbe4eced52e9c1cea7812a3428c168dfddf7cb15dbed18d561R104

File: src/add-to-project.ts [hunk 22]
```diff
@@ -152,11 +207,44 @@ export async function addToProject(): Promise<void> {
       }`,
       {
         projectId,
-        title: issue?.html_url
-      }
+        title: issue?.html_url,
+      },
```
Comment (parisk) [inline src/add-to-project.ts]: ditto: https://github.com/withlogicco/add-to-project/pull/1/files#diff-e3ec74736882b1cbe4eced52e9c1cea7812a3428c168dfddf7cb15dbed18d561R104

File: src/add-to-project.ts [hunk 23]
```diff
@@ -39,6 +59,8 @@ interface ProjectV2AddDraftIssueResponse {
 export async function addToProject(): Promise<void> {
   const projectUrl = core.getInput('project-url', {required: true})
   const ghToken = core.getInput('github-token', {required: true})
+  const fields = core.getInput('fields', {required: false})
```
Comment (parisk) [inline src/add-to-project.ts]: This should be multiline input[^1].

[^1]: https://github.com/actions/toolkit/blob/409d616a6e7610097d47caed40057fc21b43da99/packages/core/README.md?plain=1#L26

File: src/add-to-project.ts [hunk 24]
```diff
@@ -113,9 +135,41 @@ export async function addToProject(): Promise<void> {
...
+            }
+          }
+        }
+      }
+    }`,
+    {
+      projectId,
+    },
+  )
+
```
Comment (parisk) [inline src/add-to-project.ts]: Let's move this action below, along with the rest "fields" business logic[^1] and only run it if, the multiline fields input has length > 0.

[^1]: https://github.com/withlogicco/add-to-project/pull/1/files#diff-e3ec74736882b1cbe4eced52e9c1cea7812a3428c168dfddf7cb15dbed18d561R219

File: src/add-to-project.ts [hunk 25]
```diff
@@ -152,11 +207,44 @@ export async function addToProject(): Promise<void> {
...
+        title: issue?.html_url,
+      },
     )
 
     core.setOutput('itemId', addResp.addProjectV2DraftIssue.projectItem.id)
+    itemId = addResp.addProjectV2DraftIssue.projectItem.id
+  }
+
+  // check if fields is empty string
+  if (fields) {
```
Comment (parisk) [inline src/add-to-project.ts]: A couple of comments here:

- [x] Since `fields` should be a multiline string, we should check against the input's length
- [x] Let's prefer an early return, instead of nesting our logic in an `if` statement. Example
    ```js
    if (!fields.length) {
      return;
    }
    
    // our fields business logic here
    ```

File: src/check-custom-fields.ts [hunk 26]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+interface ArrayType {
+  options: OptionType[]
+  name: string
+  dataType: string
+  id: string
+}
+
+export async function checkDictionaryInArray(obj: {[key: string]: string}, arr: []) {
```
Comment (parisk) [inline src/check-custom-fields.ts]: A few comments here:

- [x] Add JSDoc documentation explaining what this function does.
- [x] Rename the function to `getFieldUpdates`
- [x] Rename argument `obj` to `fieldValues`
- [x] Create a `FieldValueType` interface for `fieldValues`
- [x] Rename `arr` to `projectFields`
- [x] Set the type of `projectFields` to it's actual type.

File: src/add-to-project.ts [hunk 27]
```diff
@@ -1,5 +1,6 @@
 import * as core from '@actions/core'
 import * as github from '@actions/github'
+import * as check from './check-custom-fields'
```
Comment (parisk) [inline src/add-to-project.ts]: Let's rename this module to something more generic, like `utils`.

```suggestion
import * as utils from './utils'
```

File: src/check-custom-fields.ts [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+          result.push({id: numberId, value: {numberValue}})
+          break
+
+        default:
+          core.debug(`Unsupported data type: ${found.dataType}`)
+          break
+      }
+    }
+  }
+  core.debug(`Result: ${JSON.stringify(result)}`)
```
Comment (parisk) [inline src/check-custom-fields.ts]: Let's remove this

```suggestion
```

File: src/check-custom-fields.ts [hunk 29]
```diff
@@ -0,0 +1,64 @@
...
+      // TODO: Add support for other data types (e.g. ITERATIONS)
+      switch (found.dataType) {
+        case 'SINGLE_SELECT':
+          const optionFound = found.options.find(option => option.name.toLowerCase() === object[key].toLowerCase())
+
+          if (optionFound) {
+            const id = found.id
+            const value = optionFound.id
+            result.push({id, value: {singleSelectOptionId: value}})
+          }
```
Comment (parisk) [inline src/check-custom-fields.ts]: Let's use `continue` here

```suggestion
          if (!optionFound) {
            // warn that the given value was not found as choice
            continue
          }

          const id = found.id
          const value = optionFound.id
          result.push({id, value: {singleSelectOptionId: value}})
```

File: src/check-custom-fields.ts [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+interface ArrayType {
+  options: OptionType[]
+  name: string
+  dataType: string
+  id: string
+}
+
+export async function checkDictionaryInArray(obj: {[key: string]: string}, arr: []) {
+  const object: ObjectType = obj
+  const array: ArrayType[] = arr
```
Comment (parisk) [inline src/check-custom-fields.ts]: Why do we need these here?

File: src/check-custom-fields.ts [hunk 31]
```diff
@@ -0,0 +1,64 @@
...
+
+  const result = []
+
+  //remove empty objects from array of objects
+  const filteredArray = array.filter(objs => !(objs && Object.keys(objs).length === 0 && objs.constructor === Object))
```
Comment (parisk) [inline src/check-custom-fields.ts]: Great. A couple of comments:

- [ ] Let's rename `filteredArray` to `clearedProjectFields`
- [ ] Let's move the filter function to a named function (e.g. `isNotEmpty`):
    ```ts
    const isNotEmpty = obj => !(obj && Object.keys(obj).length === 0 && obj.constructor === Object)
    ```

File: src/check-custom-fields.ts [hunk 32]
```diff
@@ -0,0 +1,64 @@
...
+
+  for (const key in object) {
+    const found = filteredArray.find(item => item.name.toLowerCase() === key.toLowerCase())
+
+    if (found && found.dataType) {
```
Comment (parisk) [inline src/check-custom-fields.ts]: Let's prefer a `continue` workflow here

```ts
if (!found?.dataType) {
  // warn that no field was found for the given field name
  continue
}
```

File: src/check-custom-fields.ts [hunk 33]
```diff
@@ -0,0 +1,64 @@
...
+  const object: ObjectType = obj
+  const array: ArrayType[] = arr
+
+  const result = []
+
+  //remove empty objects from array of objects
+  const filteredArray = array.filter(objs => !(objs && Object.keys(objs).length === 0 && objs.constructor === Object))
+
+  for (const key in object) {
+    const found = filteredArray.find(item => item.name.toLowerCase() === key.toLowerCase())
```
Comment (parisk) [inline src/check-custom-fields.ts]: Let's use some gorgeous JavaScript syntax to make this more legible

```suggestion
  for (const [fieldName, fieldValue] of Object.entries(fieldValues)) {
    const foundField = clearedProjectFields.find(projectField => projectField.name.toLowerCase() === fieldName.toLowerCase())
```

File: src/check-custom-fields.ts [hunk 34]
```diff
@@ -0,0 +1,64 @@
...
+          break
+
+        case 'NUMBER':
+          const numberId = found.id
+          const numberValue = found.name
+          result.push({id: numberId, value: {numberValue}})
+          break
+
+        default:
+          core.debug(`Unsupported data type: ${found.dataType}`)
```
Comment (parisk) [inline src/check-custom-fields.ts]: A couple of things here:

- [x] Convert this to a warning, since it affects user experience.
- [x] Include field name in message (`Unsupported data type for field ${fieldName}: ${found.dataType}`)

**Review by parisk — CHANGES_REQUESTED** [review]
Great job @phemmylintry. Now we want to polish the structure of this effort, so we can can contribute it upstream to GitHub, so we have plenty of comments here!

👏 👏 👏

File: README.md [hunk 35]
```diff
@@ -142,7 +142,6 @@ Using these events ensure that a given issue or pull request, in the workflow's
 
 - add the newly created PAT as a repository secret, this secret will be referenced by the [github-token input](#github-token)
   _See [Encrypted secrets](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository) for more information_
-
```
Comment (akalipetis) [inline README.md]: Please revert this change.

File: src/add-to-project.ts (hunk 19 above)
Comment (akalipetis) [reply src/add-to-project.ts]: The change is still here.

File: src/add-to-project.ts [hunk 36]
```diff
@@ -157,6 +181,76 @@ export async function addToProject(): Promise<void> {
...
+      }
+    }`,
+    {
+      projectId,
+    },
+  )
+    
+  if (!fields.length) {
+    return
+  }
```
Comment (akalipetis) [inline src/add-to-project.ts]: This should go above `fieldsResp` - we don't need this request if there are not fields.

File: src/add-to-project.ts [hunk 37]
```diff
@@ -157,6 +181,76 @@ export async function addToProject(): Promise<void> {
...
+  const projectFields = fieldsResp.node?.fields.nodes ?? []
+  
+  const fieldsObj: {[key: string]: string} = {}
+    
+  // check if fields is empty string
```
Comment (akalipetis) [inline src/add-to-project.ts]: This is probably a leftover.

File: src/utils.ts [hunk 38]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * Returns an array of field updates for the GraphQL API
+ 
+ * @param fieldValues
+ * @param projectFields
+ * @returns {Promise<[]>}
+*/
+
+export async function getFieldUpdates(fieldValues: {[key: string]: string}, projectFields: []) {
+  const array: ProjectFieldsType[] = projectFields
```
Comment (akalipetis) [inline src/utils.ts:24]: Why do we need this `const` definition here? Can't we use `projectFields`?

File: src/utils.ts [hunk 39]
```diff
@@ -0,0 +1,78 @@
...
+  for (const [fieldName, fieldValue] of Object.entries(fieldValues)) {
+    const foundField = clearedProjectFields.find(
+      projectField => projectField.name.toLowerCase() === fieldName.toLowerCase(),
+    )
+
+    if (!foundField?.dataType) {
+      core.debug(`Could not find field with name ${fieldName}`)
+      continue
+    }
+    // TODO: Add support for other data types (e.g. ITERATIONS)
```
Comment (akalipetis) [inline src/utils.ts]: Let's remove this for now.

File: src/utils.ts [hunk 40]
```diff
@@ -0,0 +1,78 @@
...
+    switch (foundField.dataType) {
+      case 'SINGLE_SELECT':
+        const optionFound = foundField.options.find(option => option.name.toLowerCase() === fieldValue.toLowerCase())
+
+        if (!optionFound) {
+          core.debug(`Could not find option for field ${foundField.name} with value ${fieldValue}`)
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
          core.warning(`Could not find option for field ${foundField.name} with value ${fieldValue}`)
```

This should be a warning.

File: src/utils.ts [hunk 41]
```diff
@@ -0,0 +1,78 @@
...
+    !(objs && Object.keys(objs).length === 0 && objs.constructor === Object)
+  const clearedProjectFields = array.filter(isNotEmpty)
+
+  for (const [fieldName, fieldValue] of Object.entries(fieldValues)) {
+    const foundField = clearedProjectFields.find(
+      projectField => projectField.name.toLowerCase() === fieldName.toLowerCase(),
+    )
+
+    if (!foundField?.dataType) {
+      core.debug(`Could not find field with name ${fieldName}`)
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
      core.warning(`Could not find field with name ${fieldName}`)
```

This is a warning.

File: src/utils.ts [hunk 42]
```diff
@@ -0,0 +1,78 @@
...
+        const optionFound = foundField.options.find(option => option.name.toLowerCase() === fieldValue.toLowerCase())
+
+        if (!optionFound) {
+          core.debug(`Could not find option for field ${foundField.name} with value ${fieldValue}`)
+          continue
+        }
+        const id = foundField.id
+        const value = optionFound.id
+        result.push({id, value: {singleSelectOptionId: value}})
+        break
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
        continue
```

File: src/utils.ts [hunk 43]
```diff
@@ -0,0 +1,78 @@
...
+
+      case 'TEXT':
+        const textId = foundField.id
+        const textValue = foundField.name
+        result.push({id: textId, value: {textValueName: textValue}})
+        break
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
        continue
```

File: src/utils.ts [hunk 44]
```diff
@@ -0,0 +1,78 @@
...
+
+      case 'NUMBER':
+        const numberId = foundField.id
+        const numberValue = foundField.name
+        result.push({id: numberId, value: {numberValue}})
+        break
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
        continue
```

File: src/utils.ts [hunk 45]
```diff
@@ -0,0 +1,78 @@
...
+
+      default:
+        core.warning(`Unsupported data type for field ${fieldName}: ${foundField.dataType}`)
+        break
```
Comment (akalipetis) [inline src/utils.ts]:
```suggestion
        continue
```

**Review by akalipetis — CHANGES_REQUESTED** [review]
We're getting super close! Added a few more comments.

File: src/utils.ts (hunk 38 above)
Comment (phemmylintry) [reply src/utils.ts:24]: we need ro declare the properties of the projectfields, or else we get an error of property name does not exist

**Review by akalipetis — APPROVED** [review]
LGTM

**Review by parisk — APPROVED** [review]
